pkg/packetfilter/packetfilter.go

```diff
@@ -0,0 +1,192 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package packetfilter
+
+import (
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/log"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var logger = log.Logger{Logger: logf.Log.WithName("Packetfilter")}
+
+const (
+	TableTypeFilter TableType = iota
+	TableTypeRoute            // mangle
+	TableTypeNAT
+	TableTypeMAX
+)
+
+type TableType uint32
+
+type GlobalEgressHandle string
+
+type RuleAction uint32
+
+const (
+	RuleActionJump RuleAction = iota
+	RuleActionAccept
+	RuleActionMss
+	RuleActionMark
+	RuleActionSNAT
+	RuleActionDNAT
+	RuleActionMAX
+)
+
+type RuleProto uint32
+
+const (
+	RuleProtoUndefined RuleProto = iota
+	RuleProtoAll
+	RuleProtoTCP
+	RuleProtoUDP
+	RuleProtoICMP
+)
+
+type MssClampType uint32
+
+const (
+	UndefinedMSS MssClampType = iota
+	ToPMTU
+	ToValue
+)
+
+type ChainHook uint32
+
+const (
+	ChainHookPrerouting ChainHook = iota
+	ChainHookInput
+	ChainHookForward
+	ChainHookOutput
+	ChainHookPostrouting
+	ChainHookMAX
+)
+
+type ChainPriority uint32
+
+const (
+	ChainPriorityFirst ChainPriority = iota
+	ChainPriorityLast
+)
+
+type ChainType uint32
+
+const (
+	ChainTypeFilter ChainType = iota
+	ChainTypeRoute            // mangle
+	ChainTypeNAT
+	ChainTypeMAX
+)
+
+type ChainPolicy uint32
+
+const (
+	ChainPolicyAccept ChainType = iota
+	ChainPolicyDrop
+)
+
+type Rule struct {
+	DestCIDR string
+	SrcCIDR  string
+
+	SrcSetName  string
+	DestSetName string
+
+	SnatCIDR string
+	DnatCIDR string
+
+	OutInterface string
+	InInterface  string
+	TargetChain  string
+	MssValue     string
+
+	DPort     string
+	MarkValue string
+	Action    RuleAction
+	Proto     RuleProto
+	ClampType MssClampType
+}
+
+// Supported policy values are accept (which is the default) or drop.
+type Chain struct {
+	Name   string
+	Policy ChainPolicy
+}
+
+type ChainIPHook struct {
+	Name     string
+	Type     ChainType
+	Hook     ChainHook
+	Priority ChainPriority
+	Policy   ChainPolicy
+	JumpRule *Rule
+}
+
+type DrvInterface interface {
+	// Chains
+	ChainExists(table TableType, chain string) (bool, error)
+	CreateIPHookChainIfNotExists(chain *ChainIPHook) error
+	CreateChainIfNotExists(table TableType, chain *Chain) error
+	DeleteIPHookChain(chain *ChainIPHook) error
+	DeleteRegularChain(table TableType, chain string) error
+	ClearChain(table TableType, chain string) error
+
+	// rules
+	Delete(table TableType, chain string, ruleSpec *Rule) error
+	AppendUnique(table TableType, chain string, ruleSpec *Rule) error
+}
+
+type Driver interface {
+	DrvInterface
+	RuleToRuleSpec(rule *Rule) ([]string, error)
+	List(table TableType, chain string) ([]string, error)
+	DeleteRuleSpec(table TableType, chain string, rulespec ...string) error
+	Append(table TableType, chain string, rulespec ...string) error
+	Insert(table TableType, chain string, pos int, rulespec ...string) error
```

Comment:
So we have some rule methods that take a generalized `Rule` and some that take a rule string. This seems inconsistent. Also the string form is implementation-specific and I don't think we should expose via the generalized interface, ie all methods should reference `Rule`.  I assume you exposed the string form for the `List` method. However, for itpables, it seems if we can convert `Rule` to the itpables string form we should be able to convert the string back to a `Rule`. So `List` would be `List(table TableType, chain string) ([]Rule, error)`. In the `Adapter`, we should be able to compare `Rule` instances. That would also obviate defining 2 driver interfaces to avoid exposing the stringified versions. The interface would be:

```
type Driver interface {
	// Chains
	ChainExists(table TableType, chain string) (bool, error)
	CreateIPHookChainIfNotExists(chain *ChainIPHook) error
	CreateChainIfNotExists(table TableType, chain *Chain) error
	DeleteIPHookChain(chain *ChainIPHook) error
	DeleteRegularChain(table TableType, chain string) error
	ClearChain(table TableType, chain string) error

	// rules
	Delete(table TableType, chain string, ruleSpec *Rule) error
	AppendUnique(table TableType, chain string, ruleSpec *Rule) error
        Append(table TableType, chain string, ruleSpec *Rule) error
	Insert(table TableType, chain string, pos int, ruleSpec *Rule)
        List(table TableType, chain string) ([]Rule, error)
}
```

Reply:
I wrote a `FromRuleSpec` function in [this commit](https://github.com/tpantelis/submariner/pull/84/commits/b144b9040e3d34f7da3793b19dfdd01319687290) with unit tests. The `Rule`s tested are taken from consumer code to cover all the rule actions.

Reply:
I copied your FromRuleSpec function (added you as a co-author )  , thanks.